README
NAME
    Sub::Rate - Rate based sub dispatcher generator

SYNOPSIS
        my $rate = Sub::Rate->new( max_rate => 100 );
        $rate->add( 10 => sub { ... } );     # sub1
        $rate->add( 20 => sub { ... } );     # sub2
        $rate->add( default => sub { ... }); # default sub
    
        my $func = $rate->generate;

        # Calling this $func then:
        # sub1 will be called by rate about 10/100 (10%),
        # sub2 will be called by rate about 20/100 (20%),
        # default sub will be called in rest case (70%),
        $func->();

DESCRIPTION
    Sub::Rate generates a SUB that will dispatch some SUBs by specified
    rate.

CLASS METHODS
  new(%options)
        my $obj = Sub::Rate->new;

    Create Sub::Rate object.

    Available options are:

    *   max_rate => 'Number'

        Max rate. (Default: 100)

    *   rand_func => 'CodeRef'

        Random calculate function. Default is:

            sub {
                CORE::rand($_[0]);
            };

        You can change random function to your own implementation by this
        option. "max_rate" is passed as $_[0] to this function.

    *   sort => 'Bool'

  METHODS
  add($rate : Number|Str, $sub :CodeRef)
        $obj->add( 10, sub { ... } );
        $obj->add( 20, sub { ... } );
        $obj->add( 'default', sub { ... } );

    Add $sub to internal sublist rate by $rate.

    If $rate is not number but "default", then $sub is registered as default
    sub. If default sub is already registered, it will be replaced.

  generate()
        my $sub = $obj->generate;

    Create a new sub that dispatch functions by its rates.

  clear()
        $obj->clear;

    Clear all registered functions and default function.

AUTHOR
    Daisuke Murase <typester@cpan.org>

COPYRIGHT AND LICENSE
    Copyright (c) 2012 KAYAC Inc. All rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

    The full text of the license can be found in the LICENSE file included
    with this module.



Changes
Revision history for Perl extension Sub::Rate

{{$NEXT}}

0.06 2020-10-20T14:53:29Z
     - Switch to Moo from Any::Moose #4

0.05 2012-12-06T14:50:11+09:00
     - fixed a bug that NoMaxRate does not support bigger max_rate than 100 (koba04)

0.04 2012-11-10T17:12:18+09:00
     - added Sub::Rate::NoMaxRate

0.03 2012-08-10T13:45:21+09:00
     - don't sort funcs list by default

0.02 2012-07-17T21:48:47+09:00
     - added args support to generated func

0.01 2012-07-17T18:05:02+09:00
     - initial version


Makefile.PL
use strict;
use warnings;

BEGIN {
    my @devmods = qw(
        Module::Install::AuthorTests
        Module::Install::ReadmeFromPod
        Module::Install::Repository
    );
    my @not_available;

    {
        $| = 1;
        local *STDOUT;
        eval qq{use inc::Module::Install; 1;} or push @not_available, 'inc::Module::Install';
    }

    for my $mod (@devmods) {
        eval qq{require $mod} or push @not_available, $mod;
    }
    if (@not_available) {
        print qq{# The following modules are not available.\n};
        print qq{# `$^X $0 | cpanm` will install them:\n};
        print $_, "\n" for @not_available;
        print "\n";
        exit -1;
    }
}

use inc::Module::Install;
$Module::Install::AUTHOR = 0 if $ENV{TRAVIS};

name 'Sub-Rate';
all_from 'lib/Sub/Rate.pm';
readme_from 'lib/Sub/Rate.pm';

author_tests 'xt';
auto_set_repository;

test_requires 'Test::More' => '0.96';

requires 'perl' => '5.008001';
requires 'Moo';

WriteAll;


lib/Sub/Rate/NoMaxRate.pm
package Sub::Rate::NoMaxRate;
use strict;
use warnings;
use Mouse;

use List::Util 'sum';

extends 'Sub::Rate';

has '+max_rate' => 'default' => 0;

before add => sub {
    my ($self, $rate) = @_;
    $self->max_rate($self->max_rate + $rate);
};

after clear => sub {
    my ($self) = @_;
    $self->max_rate(0);
};

no Mouse;

__PACKAGE__->meta->make_immutable;

__END__

=for stopwords SUBs sublist

=head1 NAME

Sub::Rate::NoMaxRate - auto calculate max_rate

=head1 SYNOPSIS

    my $rate = Sub::Rate::NoMaxRate->new;
    $rate->add( 10 => sub { ... } );     # sub1
    $rate->add( 20 => sub { ... } );     # sub2
    
    my $func = $rate->generate;

    # Calling this $func then:
    # sub1 will be called by rate about 10/30
    # sub2 will be called by rate about 20/30
    $func->();
    
=head1 DESCRIPTION

Sub::Rate::NoMaxRate is a subclass of L<Sub::Rate>.

This module has no C<max_rate> option and calculate it automatically.

=head1 CLASS METHODS

=head2 new(%options)

    my $obj = Sub::Rate->new;

Create Sub::Rate object.

Available options are:

=over

=item * rand_func => 'CodeRef'

Random calculate function. Default is:

    sub {
        CORE::rand($_[0]);
    };

You can change random function to your own implementation by this option.
C<max_rate> is passed as C<$_[0]> to this function.

=back

=head2 METHODS

=head2 add($rate : Number|Str, $sub :CodeRef)

    $obj->add( 10, sub { ... } );
    $obj->add( 20, sub { ... } );
    $obj->add( 'default', sub { ... } );

Add C<$sub> to internal sublist rate by C<$rate>.

If C<$rate> is not number but "default", then C<$sub> is registered as default sub.
If default sub is already registered, it will be replaced.

=head2 generate()

    my $sub = $obj->generate;

Create a new sub that dispatch functions by its rates.

=head2 clear()

    $obj->clear;

Clear all registered functions and default function.

=head1 AUTHOR

Daisuke Murase <typester@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2012 KAYAC Inc. All rights reserved.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut



lib/Sub/Rate.pm
package Sub::Rate;
use strict;
use warnings;
use Moo;
use Carp;

our $VERSION = '0.05';

has max_rate => (
    is      => 'rw',
    default => 100,
);

has rand_func => (
    is      => 'rw',
    default => sub {
        return sub { rand($_[0]) };
    },
);

has sort => (
    is      => 'rw',
    default => 0,
);

has _func => (
    is      => 'rw',
    default => sub { [] },
);

has _default_func => (
    is => 'rw',
);

no Moo;

sub add {
    my ($self, $rate, $func) = @_;

    if ($rate eq 'default') {
        $self->_default_func($func);
    }
    else {
        my $total_rate = 0;
        $total_rate += $_->[0] for @{ $self->_func };

        if ($total_rate + $rate > $self->max_rate) {
            croak sprintf 'Exceed max_rate, current:%s max:%s',
                $total_rate + $rate, $self->max_rate;
        }

        push @{ $self->_func }, [$rate, $func];
    }
}

sub generate {
    my ($self) = @_;

    my @sorted_funcs = @{ $self->_func };
    @sorted_funcs = CORE::sort { $a->[0] <=> $b->[0] } @sorted_funcs if $self->sort;

    my $rand         = $self->rand_func;
    my $max_rate     = $self->max_rate;
    my $default_func = $self->_default_func;

    sub {
        my @args = @_;

        my $index  = $rand->($max_rate);
        my $cursor = 0;

        for my $f (@sorted_funcs) {
            $cursor += $f->[0];

            if ($index <= $cursor) {
                return $f->[1]->(@args);
            }
        }

        if ($default_func) {
            return $default_func->(@args);
        }
        else {
            return;
        }
    };
}

sub clear {
    my ($self) = @_;
    $self->_func([]);
    $self->_default_func(undef);
}

__PACKAGE__->meta->make_immutable;

__END__

=for stopwords SUBs sublist

=head1 NAME

Sub::Rate - Rate based sub dispatcher generator

=head1 SYNOPSIS

    my $rate = Sub::Rate->new( max_rate => 100 );
    $rate->add( 10 => sub { ... } );     # sub1
    $rate->add( 20 => sub { ... } );     # sub2
    $rate->add( default => sub { ... }); # default sub
    
    my $func = $rate->generate;

    # Calling this $func then:
    # sub1 will be called by rate about 10/100 (10%),
    # sub2 will be called by rate about 20/100 (20%),
    # default sub will be called in rest case (70%),
    $func->();
    
=head1 DESCRIPTION

Sub::Rate generates a SUB that will dispatch some SUBs by specified rate.

=head1 CLASS METHODS

=head2 new(%options)

    my $obj = Sub::Rate->new;

Create Sub::Rate object.

Available options are:

=over

=item * max_rate => 'Number'

Max rate. (Default: 100)

=item * rand_func => 'CodeRef'

Random calculate function. Default is:

    sub {
        CORE::rand($_[0]);
    };

You can change random function to your own implementation by this option.
C<max_rate> is passed as C<$_[0]> to this function.

=item * sort => 'Bool'

=back

=head2 METHODS

=head2 add($rate : Number|Str, $sub :CodeRef)

    $obj->add( 10, sub { ... } );
    $obj->add( 20, sub { ... } );
    $obj->add( 'default', sub { ... } );

Add C<$sub> to internal sublist rate by C<$rate>.

If C<$rate> is not number but "default", then C<$sub> is registered as default sub.
If default sub is already registered, it will be replaced.

=head2 generate()

    my $sub = $obj->generate;

Create a new sub that dispatch functions by its rates.

=head2 clear()

    $obj->clear;

Clear all registered functions and default function.

=head1 AUTHOR

Daisuke Murase <typester@cpan.org>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2012 KAYAC Inc. All rights reserved.

This program is free software; you can redistribute
it and/or modify it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

=cut
